torch/utils/_benchmark/utils/valgrind_wrapper/timer_interface.py

```diff
@@ -0,0 +1,448 @@
+            baseline_inclusive_stats=baseline_inclusive_stats,
+            baseline_exclusive_stats=baseline_exclusive_stats,
+            stmt_inclusive_stats=stmt_inclusive_stats,
+            stmt_exclusive_stats=stmt_exclusive_stats,
+        )
+
+    def _invoke(self, stmt: str, setup: str, number: int, num_threads: int):
+        """Core invocation method for Callgrind collection.
+
+        Valgrind operates by effectively replacing the CPU with an emulated
+        version which allows it to instrument any code at the cost of severe
+        performance degradation. This has the practical effect that in order
+        to collect Callgrind statistics, a new process has to be created
+        running under `valgrind`. The steps for this process are:
+
+        1) Create a scratch directory.
+        2) Codegen a run script. (_ValgrindWrapper._construct_script)
+            Inside the run script:
+                * Validate that Python and torch match the parent process
+                * Validate that it is indeed running under valgrind
+                * Execute `setup` and warm up `stmt`
+                * Begin collecting stats
+                * Run the `stmt` loop
+                * Stop collecting stats
+        3) Parse the run results.
+        4) Cleanup the scratch directory.
+        """
+        working_dir = tempfile.mkdtemp()
+        script_file = os.path.join(working_dir, "timer_callgrind.py")
+        callgrind_out = os.path.join(working_dir, "callgrind.out")
+        error_log = os.path.join(working_dir, "error.txt")
+        stat_log = os.path.join(working_dir, "callgrind_stat.txt")
+        stdout_stderr_log = os.path.join(working_dir, "stdout_stderr.log")
+
+        def run(args, **kwargs):
+            # https://thraxil.org/users/anders/posts/2008/03/13/Subprocess-Hanging-PIPE-is-your-enemy/
+            f_stdout_stderr = open(stdout_stderr_log, "wb")
+            try:
+                invocation = subprocess.run(
+                    args,
+                    stdout=f_stdout_stderr,
+                    stderr=subprocess.STDOUT,
+                    **kwargs,
+                )
+                with open(stdout_stderr_log, "rt") as f:
+                    return invocation, f.read()
+            finally:
+                f_stdout_stderr.close()
+
+        try:
+            with open(script_file, "wt") as f:
+                f.write(self._construct_script(
+                    stmt=stmt, setup=setup, number=number,
+                    num_threads=num_threads, error_log=error_log,
+                    stat_log=stat_log))
+
+            valgrind_invocation, valgrind_invocation_output = run([
+                "valgrind",
+                "--tool=callgrind",
+                f"--callgrind-out-file={callgrind_out}",
+                "--dump-line=yes",
+                "--dump-instr=yes",
+                "--collect-jumps=yes",
+                "--instr-atstart=yes",
```

Comment:
Shouldn't `--instr-atstart=no`? This would reduce the starting overhead (you mentioned in earlier discussion that startup was really high and wouldn't this explain it?) You turn on instrumentation later with `CALLGRIND_START_INSTRUMENTATION`)

Reply:
The earlier benchmark included `--instr-atstart=no` I went back and remeasured, and the current time for a trivial measurement is 35 sec compared to the 20s just to import torch. (The earlier higher numbers may have been double counting by collecting a baseline too. Not sure.) And I'd prefer to keep the run hermetic, which `CALLGRIND_START_INSTRUMENTATION` is not.